src/VecSim/algorithms/hnsw/hnswlib.h

```diff
@@ -1068,26 +1088,73 @@ tableint HierarchicalNSW<dist_t>::searchBottomLayerEP(const void *query_data) co
+HierarchicalNSW<dist_t>::searchBottomLayer_WithTimeout(tableint ep_id, const void *data_point,
+                                                       size_t ef, size_t k, void *timeoutCtx,
+                                                       VecSimQueryResult_Code *rc) const {
+    candidatesLabelsMaxHeap<dist_t> results(this->allocator);
+
+#ifdef ENABLE_PARALLELIZATION
+    this->visited_nodes_handler =
+        this->visited_nodes_handler_pool->getAvailableVisitedNodesHandler();
+#endif
 
-    vecsim_stl::max_priority_queue<std::pair<dist_t, labeltype>> result(this->allocator);
-    if (cur_element_count == 0)
-        return result;
+    tag_t visited_tag = this->visited_nodes_handler->getFreshTag();
 
-    tableint bottom_layer_ep = searchBottomLayerEP(query_data);
-    vecsim_stl::max_priority_queue<pair<dist_t, tableint>> top_candidates =
-        searchLayer(bottom_layer_ep, query_data, 0, std::max(ef_, k));
+    candidatesMaxHeap<dist_t> top_candidates(this->allocator);
+    candidatesMaxHeap<dist_t> candidate_set(this->allocator);
+
+    dist_t dist = fstdistfunc_(data_point, getDataByInternalId(ep_id), dist_func_param_);
+    dist_t lowerBound = dist;
+    top_candidates.emplace(dist, ep_id);
+    candidate_set.emplace(-dist, ep_id);
+
+    this->visited_nodes_handler->tagNode(ep_id, visited_tag);
+
+    while (!candidate_set.empty()) {
+        std::pair<dist_t, tableint> curr_el_pair = candidate_set.top();
+        if ((-curr_el_pair.first) > lowerBound) {
+            break;
+        }
+        if (__builtin_expect(VecSimIndex::timeoutCallback(timeoutCtx), 0)) {
+            *rc = VecSim_QueryResult_TimedOut;
+            return results;
+        }
+        candidate_set.pop();
 
+        lowerBound = process_candidate(curr_el_pair.second, data_point, 0, ef, visited_tag,
+                                       top_candidates, candidate_set, lowerBound);
+    }
+#ifdef ENABLE_PARALLELIZATION
+    visited_nodes_handler_pool->returnVisitedNodesHandlerToPool(this->visited_nodes_handler);
+#endif
     while (top_candidates.size() > k) {
         top_candidates.pop();
     }
     while (top_candidates.size() > 0) {
-        std::pair<dist_t, tableint> rez = top_candidates.top();
-        result.push(std::pair<dist_t, labeltype>(rez.first, getExternalLabel(rez.second)));
+        auto &res = top_candidates.top();
+        results.emplace(res.first, getExternalLabel(res.second));
         top_candidates.pop();
     }
-    return result;
+    *rc = VecSim_QueryResult_OK;
+    return results;
+}
+
+template <typename dist_t>
+candidatesLabelsMaxHeap<dist_t>
+HierarchicalNSW<dist_t>::searchKnn(const void *query_data, size_t k, void *timeoutCtx,
+                                   VecSimQueryResult_Code *rc) const {
+
+    if (cur_element_count == 0) {
+        *rc = VecSim_QueryResult_OK;
+        return candidatesLabelsMaxHeap<dist_t>(this->allocator);
+    }
+
+    tableint bottom_layer_ep = searchBottomLayerEP(query_data);
```

Comment:
don't we want to timeout on this also?

Reply:
By looking at profile stacks, this phase takes less then 5% of searchKnn() time, so I don't think we have to, but I can add it as well